File: 2024 April/Sum of Products/main.py
# Link: https://www.geeksforgeeks.org/problems/sum-of-products5049/1


class Solution:
    def pairAndSum(self, n, arr):
        bit_counts, bit_and_sum = [0] * 32, 0
        for a in arr:
            for i in range(32):
                mask = 1 << i
                if a & mask:
                    bit_and_sum += bit_counts[i] * mask
                    bit_counts[i] += 1
                elif mask > a:
                    break
        return bit_and_sum


if __name__ == "__main__":
    obj = Solution()
    arr = [5, 10, 15]
    n = len(arr)
    print(obj.pairAndSum(n, arr))
    arr = [10, 20, 30, 40]
    n = len(arr)
    print(obj.pairAndSum(n, arr))

File: 2024 April/Sum of Products/main.cpp
// Link: https://www.geeksforgeeks.org/problems/sum-of-products5049/1

#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    long long pairAndSum(int n, long long arr[])
    {
        vector<int> cnt(32, 0);
        for (int j = 0; j < 32; j++)
        {
            for (int i = n - 1; i >= 0; i--)
            {
                if (arr[i] & (1 << j))
                    cnt[j]++;
            }
        }
        long long int ans = 0;
        for (int j = 0; j < 32; j++)
        {
            long long val = 0;
            for (int i = 0; i < n - 1; i++)
            {
                if (arr[i] & (1 << j))
                {
                    cnt[j]--;
                    if (cnt[j] > 0)
                        val += (1LL << j) * cnt[j];
                }
            }
            ans += val;
        }
        return ans;
    }
};

int main()
{
    Solution obj;
    long long arr1[] = {5, 10, 15};
    int n1 = sizeof(arr1) / sizeof(arr1[0]);
    cout << obj.pairAndSum(n1, arr1) << endl;
    long long arr2[] = {10, 20, 30, 40};
    int n2 = sizeof(arr2) / sizeof(arr2[0]);
    cout << obj.pairAndSum(n2, arr2) << endl;
    return 0;
}